includes/Checker/Checks/Plugin_Updater_Check.php

```diff
@@ -0,0 +1,226 @@
+
+namespace WordPress\Plugin_Check\Checker\Checks;
+
+use Exception;
+use WordPress\Plugin_Check\Checker\Check_Result;
+
+/**
+ * Check to detect plugin updater.
+ *
+ * @since n.e.x.t
+ */
+class Plugin_Updater_Check extends Abstract_File_Check {
+
+	const TYPE_PLUGIN_UPDATE_URI_HEADER = 1;
+	const TYPE_PLUGIN_UPDATER_FILE      = 3;
+	const TYPE_PLUGIN_UPDATERS          = 5;
+	const TYPE_PLUGIN_UPDATER_ROUTINES  = 7;
+	const TYPE_ALL                      = 9; // Same as all of the above with bitwise OR.
+
+	/**
+	 * Bitwise flags to control check behavior.
+	 *
+	 * @since n.e.x.t
+	 * @var int
+	 */
+	protected $flags = 0;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $flags Bitwise flags to control check behavior.
+	 */
+	public function __construct( $flags = self::TYPE_ALL ) {
+		$this->flags = $flags;
+	}
+
+	/**
+	 * Amends the given result by running the check on the given list of files.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 * @param array        $files  List of absolute file paths.
+	 *
+	 * @throws Exception Thrown when the check fails with a critical error (unrelated to any errors detected as part of
+	 *                   the check).
+	 */
+	protected function check_files( Check_Result $result, array $files ) {
+		$php_files = self::filter_files_by_extension( $files, 'php' );
+
+		// Looks for "UpdateURI" in plugin header.
+		if ( $this->flags & self::TYPE_PLUGIN_UPDATE_URI_HEADER ) {
+			$this->look_for_update_uri_header( $result );
+		}
+
+		// Looks for special updater file.
+		if ( $this->flags & self::TYPE_PLUGIN_UPDATER_FILE ) {
+			$this->look_for_updater_file( $result, $php_files );
+		}
+
+		// Looks for plugin updater code in plugin files.
+		if ( $this->flags & self::TYPE_PLUGIN_UPDATERS ) {
+			$this->look_for_plugin_updaters( $result, $php_files );
+		}
+
+		// Looks for plugin updater routines in plugin files.
+		if ( $this->flags & self::TYPE_PLUGIN_UPDATER_ROUTINES ) {
+			$this->look_for_updater_routines( $result, $php_files );
+		}
+	}
+
+	/**
+	 * Looks for UpdateURI in plugin header and amends the given result with an error if found.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 */
+	protected function look_for_update_uri_header( Check_Result $result ) {
+		if ( ! function_exists( 'get_plugin_data' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/plugin.php';
+		}
+
+		$plugin_main_file = WP_PLUGIN_DIR . '/' . $result->plugin()->basename();
+		$plugin_header    = get_plugin_data( $plugin_main_file );
+		if ( ! empty( $plugin_header['UpdateURI'] ) ) {
+			$this->add_result_error_for_file(
+				$result,
+				true,
+				__( 'Plugin Updater detected. Use of the Update URI header is not helpful in plugins hosted on WordPress.org.', 'plugin-check' ),
+				'plugin_updater_detected',
+				$plugin_main_file
+			);
+		}
```

Comment:
How do you know if this is hosted in WordPress.org?

Reply:
@spacedmonkey This test is specifically about plugin review team requirements for plugins hosted on wordpress.org, see https://github.com/WordPress/plugin-check/blob/trunk/checks/plugin-updaters.php. So we should match their requirements for that.

Reply:
That is a confusing check then. What if you run this test again your own plugin that is not hosted on WP.org repo? This is flagging a warning for something that is not an issue.

Reply:
Not every check has to be run by everyone. This here specifically is a "wordpress.org minimum requirements" check, so it also needs to be available in our plugin, otherwise developers won't be able to sanity-check their plugin for those requirements.

If someone develops a plugin for somewhere outside wordpress.org _and_ is building their own updater, they can always choose to not run this check.

Reply:
Thanks @felixarntz for the answers.